extensions/notebook/src/book/bookTocManager.ts

```diff
@@ -0,0 +1,136 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as path from 'path';
+import { BookTreeItem } from './bookTreeItem';
+import * as yaml from 'js-yaml';
+import * as fs from 'fs';
+import { IJupyterBookSectionV1, IJupyterBookSectionV2, JupyterBookSection } from '../contracts/content';
+import { BookVersion } from './bookModel';
+import * as vscode from 'vscode';
+
+export interface IBookTocManager {
+	updateBook(section: BookTreeItem, updatedBook: BookTreeItem): Promise<void>;
+	createBook(bookContentPath: string, contentFolder: string): Promise<void>
+}
+const allowedFileExtensions: string[] = ['.md', '.ipynb'];
+
+export enum tocSectionOperation {
+	Remove,
+	Add
+}
+
+export class BookTocManager implements IBookTocManager {
+
+	constructor() {
+	}
+
+	async getAllFiles(toc: IJupyterBookSectionV2[], directory: string, filesInDir: string[]): Promise<IJupyterBookSectionV2[]> {
+		await Promise.all(filesInDir.map(async file => {
+			let isDirectory = (await fs.promises.stat(path.join(directory, file))).isDirectory();
+			if (isDirectory) {
+				let jupyterSection: IJupyterBookSectionV2 = {
+					title: file,
+					file: file,
+					expand_sections: true,
+					numbered: false,
+					sections: []
+				};
+				let files = await fs.promises.readdir(path.join(directory, file));
+				toc.push(jupyterSection);
+				await this.getAllFiles(toc, path.join(directory, file), files);
+			} else if (allowedFileExtensions.includes(path.extname(file))) {
+				const addFile: IJupyterBookSectionV2 = {
+					title: path.parse(file).name,
+					file: path.join(path.basename(directory), path.parse(file).name)
+				};
+				if (toc.length > 0) {
+					let indexToc = toc.findIndex(parent => parent.title === path.dirname(addFile.file));
+					if (indexToc !== -1) {
+						if (path.parse(addFile.file).ext === '.md') {
+							toc[indexToc].file = addFile.file;
+						} else {
+							toc[indexToc].sections.push(addFile);
+						}
+					}
+				} else if (path.parse(file).ext === '.md') {
+					//TODO : Add other types such as Intro, Introduction, Index etc etc
+					// what about two files named as Intro and then other as readme ?
+					let index: IJupyterBookSectionV2 = {
+						title: path.parse(file).name,
+						file: path.parse(file).name
+					};
+					toc.push(index);
+				}
+			}
+		}));
+		return toc;
+	}
+
+	hasSections(node: JupyterBookSection): boolean {
+		return (typeof node === 'object') && node.sections.length > 0 && (typeof node.sections !== 'undefined');
```

Comment:
Seems odd to be checking that this is an object given that the type is JupyerBookSection so should always be an object